### ISSUE TITLE
UDP support

### ISSUE THREAD

**Issue by pyamsoft** [pyamsoft/tetherfusenet#101]
We may need to build a SOCKS5 proxy for UDP traffic since we can't transparent proxy UDP from an HTTP proxy

**Comment by zellchristensen** [comment]
I ran in to an issue with some UDP applications not working and would love to see it implemented.

**Comment by pyamsoft** [comment]
UDP is much trickier to proxy than TCP sadly, but this actively affects me too so its top of mind (though sadly very difficult)

**Comment by pyamsoft** [comment]
Socks with ktor
https://github.com/torsm/ktor-socks/tree/master/src/main/kotlin/de/torsm/socks

Socks UDP assoc code
https://github.com/bbottema/java-socks-proxy-server/blob/master/src/main/java/org/bbottema/javasocksproxyserver/Socks5Impl.java

**Comment by pyamsoft** [comment]
This will need to be an HTTP based solution since consoles and phones dont support socks natively.

Generally speaking you can't http proxy UDP, but we will see if we can implement a proxy over http via quic or something

**Comment by Arr4y** [comment]
I would be happy for that to happen. I've tried to use another app for hosting a different proxy while using this one to host wifi. Telegram never let's me start a voice call because of errors. I couldn't use HTTP because of UDP and socks didn't work either. Shadowsocks worked but I wasn't sure why it did and socks5 didn't.

I'm looking forward to this solution.

**Comment by Arr4y** [comment]
> This will need to be an HTTP based solution since consoles and phones dont support socks natively.
> 
> Generally speaking you can't http proxy UDP, but we will see if we can implement a proxy over http via quic or something

Perhaps a library like this could help development.
https://github.com/VKCOM/KNet

**Comment by Arr4y** [comment]
> This will need to be an HTTP based solution since consoles and phones dont support socks natively.
> 
> Generally speaking you can't http proxy UDP, but we will see if we can implement a proxy over http via quic or something

I look forward to this as I've been struggling getting some things to work. Telegram wouldn't start a call without UDP suport. I've tried some methods to get it to work. Making it work natively would make it much easier to work with.

Perhaps a library like this could help development.
https://github.com/VKCOM/KNet

**Comment by pyamsoft** [comment]
Hi,

I'm not sure about directly using this knet library, as it pulls in a ton of stuff and hasn't been touched code wise for a few years.

I appreciate the enthusiasm around the topic. I have 2 bare minimum requirements for a valid UDP solution in Tetherfi, it needs to run on a non rooted Android phone, and it needs to be possible without becoming the System VPN service or needing the system Mobile Hotspot.

UDP is a lot trickier since there is no method on Android to capture any UDP traffic like there is for setting a proxy up via the wifi settings for TCP currently.

Basically, the calling app needs to speak QUIC already, and assuming it does, possibly the existing TCP proxy might already work as its a simple forwarder from a client to a server.

If the app like Telegram only talks in UDP, I am not able to capture its socket connections without being a VPN app.

Tetherfi is not a VPN app, and it never will be. The google play store would never accept it.

UDP proxying will take time, this is assuming it is even technically possible in the first place in a way that doesnt require other apps to specifically change.

**Comment by Arr4y** [comment]
> Hi,
> 
> I'm not sure about directly using this knet library, as it pulls in a ton of stuff and hasn't been touched code wise for a few years.
> 
> I appreciate the enthusiasm around the topic. I have 2 bare minimum requirements for a valid UDP solution in Tetherfi, it needs to run on a non rooted Android phone, and it needs to be possible without becoming the System VPN service or needing the system Mobile Hotspot.
> 
> UDP is a lot trickier since there is no method on Android to capture any UDP traffic like there is for setting a proxy up via the wifi settings for TCP currently.
> 
> Basically, the calling app needs to speak QUIC already, and assuming it does, possibly the existing TCP proxy might already work as its a simple forwarder from a client to a server.
> 
> If the app like Telegram only talks in UDP, I am not able to capture its socket connections without being a VPN app.
> 
> Tetherfi is not a VPN app, and it never will be. The google play store would never accept it.
> 
> UDP proxying will take time, this is assuming it is even technically possible in the first place in a way that doesnt require other apps to specifically change.

I've used an app called "Android Proxy Server" and created a shadowsocks proxy with it. I've used it with Tether Fi to connect a laptop to it. That fixed the problem but I would have many instances where the app would stop working until I opened it again. The app doesn't pose as a VPN, it just creates a server. I've used others that created a socks proxy but they failed and wasn't sure why. But that was my current solution.

**Comment by pyamsoft** [comment]
I will look at seeing if I can implement a shadowsocks proxy in tetherfi, to avoid the current issues with running 2 apps. It could be a workable solution.

(I believe the disconnect issues are caused by both attempting to use Wifi direct to setup local hotspots. Tetherfi expects to be the exclusive Wifi direct owner, and I imagine so does the Android Proxy Server app, which is why you are probably seeing the issue.)

**Comment by Arr4y** [comment]
> I will look at seeing if I can implement a shadowsocks proxy in tetherfi, to avoid the current issues with running 2 apps. It could be a workable solution.
> 
> (I believe the disconnect issues are caused by both attempting to use Wifi direct to setup local hotspots. Tetherfi expects to be the exclusive Wifi direct owner, and I imagine so does the Android Proxy Server app, which is why you are probably seeing the issue.)

The other app does not create a wifi direct nor connection. It only hosts a server. It appears to stop working when my phone shuts off so that's most likely the app's problem.

Do we have an ETA for a shadowsocks proxy within the app?

**Comment by pyamsoft** [comment]
No ETA at the moment - I still need to see if a shadowsocks proxy would be the right path forward.

It looks like a shadowsocks server is a huge undertaking - and it would require connecting clients to run a shadowsocks client. If that is the case, it won't be a solution for the long-term as would not be supported by simpler devices - though it may be an interesting stop-gap solution.

Long term would probably look at something like CONNECT-UDP, which would supposedly be supported on anything that speaks HTTP, which is exactly what we would be looking for

https://datatracker.ietf.org/doc/html/rfc9298

**Comment by chromer030** [comment]
+1 for Socks5 proxy server.

**Comment by brlin-tw** [comment]
Note that, as a workaround, it is possible to run an SSH service on the phone(e.g. SimpleSSHD), and use OpenSSH's application-level port forwarding feature to run a SOCKS server on the localhost that forwards the requests to the phone.

This may not work with all UDP use cases, though.

**Comment by QZLin** [comment]
> Note that, as a workaround, it is possible to run an SSH service on the phone(e.g. SimpleSSHD), and use OpenSSH's application-level port forwarding feature to run a SOCKS server on the localhost that forwards the requests to the phone.
> 
> I have implemented a script to start such a service on GNU+Linux operating system: [enable-ssh-socks-service.sh.gz](https://github.com/pyamsoft/tetherfi/files/15247595/enable-ssh-socks-service.sh.gz)
> 
> This may not work with all UDP use cases, though.

Now I manually start a socks5 proxy server(3proxy) in termux when using tetherfi,  it works fine and just a bit complex.

**Comment by pyamsoft** [comment]
Good to hear that this works as a workaround for client devices that have access to a UNIXlike shell.

I assume, were I to incorporate something like this in-app, that TetherFi would launch a SOCKS SSH server on a port (2222) at the Wi-Fi direct IP (192.168.49.1), and the user would have to manually configure their client device (laptop, desktop) to point at it as a SOCKS proxy or use proxychains or something of the sort?

At that point, I would have to build TetherFi as a full SOCKS proxy, or a SOCKS and HTTP proxy - this may increase battery usage.

This doesn't sound like it would be a solution for things like gaming consoles, which do not support SOCKS, but it certainly is a step in the right direction!

**Comment by moraroy** [comment]
@pyamsoft yes implementing this would be really cool im trying to get steam to work with all these setups mentioned here in the comments but so far not working for me :/ definitely needs udp

**Comment by chromer030** [comment]
+1 for Socks5 proxy.

**Comment by moraroy** [comment]
@chromer030 could this be used as a workaround too? https://github.com/jpillora/chisel

**Comment by chromer030** [comment]
@moraroy, internal server on same port is best solution.
(Apps i know using this: nekobox, husi)

**Comment by Clyde6790pGIT** [comment]
Any updates? I'm going on a field trip for school in like a few months and I'm in charge of being the hotspot person and some people use apps that use udp

**Comment by pyamsoft** [comment]
A small progress update.

A development version of TetherFi has support for a SOCKS4 and SOCKS5 proxy on port 8229. It is not ready yet for general usage.

We generally require 4A and 5H because a device connected to TetherFi is assumed to not have another functioning Internet connection, so it must resolve DNS on the phone, though there is also support for normal 4 and 5. Future work will validate socks5 connecting binding and support for UDP association.

The code is not stable enough yet for release, so it will not come in the immediate next release (forecasted 48), but once it stabilizes, TetherFi will be a dual http and socks proxy application.

I will add the option to run either or both proxies in the future.

Even with socks support, your client device will need to support socks, via something like native support (eg firefox) or forced (eg proxychains), so game consoles like the Switch and PS4 which lack SOCKS natively won't be supported without a Router level proxy of some sort, which is out of scope.

Not a perfect solution, but definitely progress in the right direction :)

**Comment by Clyde6790pGIT** [comment]
> A small progress update.
> 
> A development version of TetherFi has support for a SOCKS4 and SOCKS5 proxy on port 8229. It is not ready yet for general usage.
> 
> We generally require 4A and 5H because a device connected to TetherFi is assumed to not have another functioning Internet connection, so it must resolve DNS on the phone, though there is also support for normal 4 and 5. Future work will validate socks5 connecting binding and support for UDP association.
> 
> The code is not stable enough yet for release, so it will not come in the immediate next release (forecasted 48), but once it stabilizes, TetherFi will be a dual http and socks proxy application.
> 
> I will add the option to run either or both proxies in the future.
> 
> Even with socks support, your client device will need to support socks, via something like native support (eg firefox) or forced (eg proxychains), so game consoles like the Switch and PS4 which lack SOCKS natively won't be supported without a Router level proxy of some sort, which is out of scope.
> 
> Not a perfect solution, but definitely progress in the right direction :)

Nice to see that we have hope on this!

**Comment by kaovilai** [comment]
I would like to add that I am currently using https://play.google.com/store/apps/details?id=cn.adonet.proxy.server.pro&hl=en-US for sock5 proxy and it works through phone's native hotspot and also via TetherFi. Claims to forward UDP traffic and very little thing don't work for me.

I was also able to on my mac connect to a tailscale exit node and other VPNs which IIUC effectively renders all services usable.

**Comment by pyamsoft** [comment]
Hello, it's been a little while. I hope everyone has had an enjoyable new years, if you celebrate, and is finding the start to 2025 to be a happy one.


I have finally managed to find some time after my holidays and a busy life schedule to release version `53` of TetherFi, which includes support for SOCKS4A and SOCKS5H proxies as an Experimental Setting.


tl;dr You can test SOCKS in the app if you are brave enough. Things may break. I have not been able to test everything myself even. Most of the code is written copying other projects and looking at RFC documents.

----

**By default, this feature is not enabled** as it can be buggy and may potentially break things. If you are brave enough however, you may enable this toggle by going to the Settings dialog.

In the Settings dialog, if you scroll down to the bottom, there is a `Danger Zone` section which includes a `Debug Mode` option. You'll need to enable this `Debug Mode` switch, and then click on the `Debug Mode` list item, which will open the `Debug Logging` dialog.

In this `Debug Logging` dialog you will see a section for experiments. As long as your TetherFi hotspot is not running, you will be able to check the box enabling the experimental SOCKS proxy.

----

Things may break, things may work. For the time being please do not expect this to be a fully usable feature. Simple things like using Firefox or curl should work (they work on my machine 😄 ) but currently I do not have a reliable and reproducible environment to test, so the code there is written to spec but completely untested 🤷 Depending on how much you understand about networking and SOCKS, this may make sense to you: `connect` is lightly tested and working on my machine, `bind` and `udp_associate` are not tested.

Currently, enabling the experimental option will turn on both HTTP and SOCKS proxies, there is no way to turn off the HTTP proxy yet.

If you do notice problems, feel free to open issues. I can't guarantee I will be able to test everything, but I obviously want to get this out the door now that the codebase has finally been progressing after having this issue open for so long 😄 Hopefully this year 😅 

Due to the fact that this feature is larger than originally planned, there is no currently set release date for the SOCKS proxy. The experimental option will stay until general availability, and I intend to support it moving forward until then.

**Comment by km-nisei** [comment]
Something worth looking into is a desktop app that pairs to the Android app. I'm a long time user of PDANet+, and everything works flawlessly in Windows through that method, save for a few online games, but that's easily fixed with a VPN on my desktop. The only real downside is it doesn't have a desktop-side app for the Linux platform. I would ditch Windows in a heartbeat if you had one.

**Comment by killerkalamari** [comment]
Working great, thanks a lot! In case it's useful to anyone, here's the contents of a [proxy-auto config](https://en.wikipedia.org/wiki/Proxy_auto-config) file that works in Firefox for the "Automatic proxy configuration URL" (to create the needed URL, prepend `file://` to an absolute path, like `file:///usr/local/share/pac.js`):
```
function FindProxyForURL(url, host) {
    return isInNet(myIpAddress(), "192.168.49.1", "255.255.255.255") ? "SOCKS5 192.168.49.1:8229" : "DIRECT";
}
```

**Comment by pyamsoft** [comment]
Regarding USB tethering and Steam

I think that once the USB is connected, data is still flowing over the carrier connection. For example, things like Steam do NOT work currently with TetherFi (steam dislikes proxies), so Steam itself used your normal data it seems.

So basically, you actually had two connections over USB, your normal carrier data and the TetherFi hotspot. Your web browser used TetherFi but Steam used your carrier data.

Unfortunately, only the Android system has full control over the USB tether connection, so there isn't much I can do to prevent this. I will add warnings in a future version about this behavior.

----

Even though you configure the HTTP/SOCKS proxy, individual apps on the system can choose to ignore it or respect it. The browser usually respects things, but Steam specifically does not respect any proxies and will always attempt a direct data connection.

If possible I would recommend that you generally use the Wi-Fi direct method in the future, USB tethering is not actually any faster, and Wi-Fi direct has the advantage that things only work over the proxy.

So in this situation, since Steam was not actually using TetherFi, it would have refused to open and thus would not have used your hotspot data.

I will continue to support the USB tethering option as best I can, but I do not personally recommend it as it is less stable and less tested. I myself do not personally use it, except when testing TetherFi. For day to day usage, I use the Wi-Fi direct method because it does not mix the system hotspot and TetherFi in any way.

**Comment by Zelar00** [comment]
Thanks

**Comment by 5igh5** [comment]
> We may need to build a SOCKS5 proxy for UDP traffic since we can't transparent proxy UDP from an HTTP proxy

do you have an update with this? I'd try to help if you'd be up for it...

**Comment by billowingbaleofbovinefodder** [comment]
I just want to jump in here to mention that I have every single aspect of Steam working properly when using OpenVPN with a protonvpn tcp configuration. I wasn't able to use a proxy (to connect from windows to to my phone using USB tethering with tetherfusenet) with the protonvpn app since it doesn't support using a proxy but using openvpn gui with a protonvpn configuration to a proxy IP on the usb phone works perfectly, I'm using it here.

I would assume this works on other VPN software that allows a proxy address. I also have some weird firewall configurations that might not be recommended that might impact something working in case no one else has this work for them. I'm fairly certain I just used the http proxy and not the socks5 but I can double check soon.

**Comment by pyamsoft** [comment]
Version 67 ( hand-up, hand-down motion) has been released to the store which contains the first iteration of the new Netty based engine.

Again, this bumps the minimum required Android OS version from N to O (which is a decade old at this point). Please test the new netty engine as it provides better, faster, more full support for both HTTP/HTTPS and SOCKS/UDP based connections scenarios.

Again, in the next version the new engine will become the default, and then next-next will become the only supported engine moving forward. At that point, I believe this ticket can be closed. Sadly TetherFuseNet itself can NEVER support a full UDP connection scheme since the technology that allows this kind of tunneling is just not standard yet. Solutions that involve multiple tools will always exist for things like proxybridge and VPN connections. TetherFuseNet has no plans to implement these kinds of solutions within itself as I would rather let these successful projects do their job, while this project should focus on facilitating your connection TO those projects.

Unless plans change in the future, and everything is vibed, and it becomes so easy to just vibe every new feature and the kitchen sink into a project, that seems to be the direction I want to take. This ticket has been open for long enough, and I have other interesting things I want to get TetherFi to do around its setup and UX.

Thanks for your continued patience, I hope you have a nice weekend.

**Comment by Jgr9** [comment]
It might have already been mentioned, but I'm happy Steam is working now, but my Steam download still seems to be getting throttled to below 600Kbps, meanwhile I can browse and Speed Test fine with hundreds of Mbps.

Is it possible to do something about it?

Oh I see from last year Steam can function over USB, but Downloads won't go around throttling?

Edit: Whoops, I just read your post mentioning Proxybridge again. I set up a Process Rule (along with TetherFuseNet) and it looks like no more Steam Throttling!
Edit2: Hmm, Steam downloads don't seem to be cooperating today, among other things, but I can browse :(

**Comment by pyamsoft** [comment]
Version 69 (nice) has entered store beta testing and with it brings the current phase of this long ticket to a close.

UDP support via SOCKS5-UDP-CONNECT is supported (we are a client), though there is admittedly only a small number of servers that support SOCKS5 UDP, they do exist.

This is not the nice "out of the box" solution I was hoping for regarding things like phones and game consoles, but it has been long enough and it will have to do for now. Perfect is the enemy of good.

As always, the march forward continues with finding cool solutions for weird Internet situations. I plan to leave this ticket pinned as I think there are useful community inputs around how to get "full support" on devices, and I find issues more useful than discussions or a static wiki, though I will link the Wiki to this issue and try to summarize the salient bits on a static page :)

**Comment by moraroy** [comment]
@Jgr9 could you share on how one would go about setting that up?
